sys/usb/usbus/hid/hid.c

```diff
@@ -0,0 +1,228 @@
+    },
+    .len_type = USBUS_DESCR_LEN_FIXED
+};
+
+static size_t _gen_hid_descriptor(usbus_t *usbus, void *arg)
+{
+    usbus_hid_device_t *hid_dev = arg;
+    usb_desc_hid_t hid_desc;
+
+    hid_desc.length = sizeof(usb_desc_hid_t);
+    hid_desc.desc_type = USB_HID_DESCR_HID;
+    hid_desc.bcd_hid = USB_HID_VERSION_BCD;
+    hid_desc.country_code = USB_HID_COUNTRY_CODE_NOTSUPPORTED;
+    hid_desc.num_descrs = 0x01;
+    hid_desc.report_type = USB_HID_DESCR_REPORT;
+    hid_desc.report_length = hid_dev->report_desc_size;
+
+    usbus_control_slicer_put_bytes(usbus, (uint8_t *)&hid_desc,
+                                   sizeof(hid_desc));
+    return sizeof(usb_desc_hid_t);
+}
+
+static void _handle_tx_ready(event_t *ev)
+{
+    usbus_hid_device_t *hid = container_of(ev, usbus_hid_device_t, tx_ready);
+
+    /* todo: use mutex to lock hid tx ? */
+
+    /* linked list ep at index 1 is IN */
+    usbdev_ep_ready(hid->iface.ep->next->ep, hid->occupied);
+}
+
+void usbus_hid_init(usbus_t *usbus, usbus_hid_device_t *hid, usbus_hid_cb_t cb,
+                    uint8_t *report_desc, size_t report_desc_size)
+{
+    memset(hid, 0, sizeof(usbus_hid_device_t));
+    hid->usbus = usbus;
+    mutex_init(&hid->in_lock);
+    hid->handler_ctrl.driver = &hid_driver;
+    hid->report_desc = report_desc;
+    hid->report_desc_size = report_desc_size;
+    hid->cb = cb;
+
+    DEBUG("hid_init: %d %d \n", report_desc_size, report_desc[0]);
+    usbus_register_event_handler(usbus, &hid->handler_ctrl);
+}
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler)
+{
+    DEBUG("USB_HID: initialization\n");
+    usbus_hid_device_t *hid = (usbus_hid_device_t *)handler;
+
+    hid->tx_ready.handler = _handle_tx_ready;
+
+    hid->hid_descr.next = NULL;
+    hid->hid_descr.funcs = &_hid_descriptor;
+    hid->hid_descr.arg = hid;
+
+    /*
+    Configure Interface as USB_HID interface, choosing NONE for subclass and
+    protocol in order to represent a generic I/O device
+    */
+    hid->iface.class = USB_CLASS_HID;
+    hid->iface.subclass = USB_HID_SUBCLASS_NONE;
+    hid->iface.protocol = USB_HID_PROTOCOL_NONE;
+    hid->iface.descr_gen = &hid->hid_descr;
+    hid->iface.handler = handler;
+
+    usbus_endpoint_t *ep = usbus_add_endpoint(usbus, &hid->iface,
+                                              USB_EP_TYPE_INTERRUPT,
+                                              USB_EP_DIR_IN,
+                                              CONFIG_USBUS_HID_INTERRUPT_EP_SIZE);
+
+    /* interrupt endpoint polling rate */
+    ep->interval = 0x05;
+    usbus_enable_endpoint(ep);
+
+    ep = usbus_add_endpoint(usbus, &hid->iface,
+                            USB_EP_TYPE_INTERRUPT, USB_EP_DIR_OUT,
+                            CONFIG_USBUS_HID_INTERRUPT_EP_SIZE);
+
+    /* interrupt endpoint polling rate */
+    ep->interval = 0x05;
+    usbus_enable_endpoint(ep);
+
+    /* signal that INTERRUPT OUT ready to receive data */
+    usbdev_ep_ready(ep->ep, 0);
```

Comment:
Why are there two endpoints?

Reply:
This is due to the fact, that this USB HID interface was needed as base for the implementation of the CTAP2 protocol (part of the FIDO2 standard) in RIOT.
CTAP2 is mostly used for security keys, which need a OUT endpoint. 

For a keyboard, the OUT endpoint is useless. There is more code needed to analyze the HID report descriptor and 
add endpoints based on that.